core: introduce extensions

AaronO opened the pull request:
This PR implements ideas discussed in https://github.com/denoland/deno/issues/9738, specifically the concept of `JsRuntimeModule`s which makes modular runtime features plug-n-play.

## Goals

Reduce complexity in `runtime`, make modules self-contained and streamline snapshotting, etc...

This should make many parts of `deno` simpler and reusable by third-parties and reduce the implicit complexity in bootstrapping JsRuntimes (when snapshotting or not ...). `op_crates` should now be easily usable without `runtime`

## Subtasks

- [x] Core `JsRuntimeModule` foundations/traits (`core/runtime_modules.rs`)
- [x] Implement `JsRuntime` init via modules
- [x] Refactor all `op_crates` to be `::init() -> JsRuntimeModule`
      (essentially move code from `runtime/ops/*.rs` to `op_crates/*/lib.rs`)
- [x] Introduce `include_js_files!()` and `declare_ops!()` macros
- [x] Refactor module consumers (places where we init `JsRuntime`s & ops to use these new modules)
- [x] Implement `OpMiddleware`
  - [x] `OpMetrics` (replacing `op_metrics`)
  - [ ] Maybe implement `OpTracing` middleware (`--optrace` flag)
  - [ ] Maybe turn the `UnstableChecker` into an OpMiddleware
- [ ] Make `deno_runtime` use JsRuntimeModules ? 
- [ ] Explore declarative plugins ? (a subset of the JsRuntimeModule interface)
- [ ] Review new abstractions (naming, fit, etc...)
- [ ] Maybe add a `web_modules()` helper function that inits all Web-API modules (url, fetch, ... essentially `op_crates` today)
- [ ] Maybe add a `--no-snapshot` flag to allow running without snapshots (not sure if useful at all, even when debugging, besides https://github.com/denoland/deno/pull/9713)

## Follow-ups

- [ ] Restore extension builder pattern (more user friendly)
- [ ] Restore `declare_ops!()` macros
- [ ] Restore `MultiModule` abstraction, thus reverting Extension to a Trait
- [ ] Use extensions in `runtime`
- [ ] Use extensions in `cli`
- [ ] Allow passing extra extensions to `Worker` and `WebWorker`, making them more reusable and composable
- [ ] UnstableChecker OpMiddleware for fully unstable ops
- [ ] Tracing OpMiddleware for debugging
- [ ] Plugins (might only be a subset of the full trait)  ...
- [ ] Allow running `Worker` and `MainWorker` without snapshots, good for debugging and in general snapshots should be optional and viewed as a graceful enhancement (perf boost)

ry commented on core/runtime_modules.rs:
````diff
@@ -0,0 +1,170 @@
+// ```
+// fn web_modules(args: WebModuleArgs) -> MultiModule {
+//  MultiModule::new(vec![deno_url::init(), deno_fetch::init(...), ...])
+// }
````
Interesting!

ry commented:
Nice! I fully support this effort. 

If you can get the existing op crates working and the tests pass, I'm rather confident that this will generalize to future needs. I think it will be challenging tho.

ry commented on runtime/build.rs:
```diff
@@ -44,8 +40,28 @@ fn create_snapshot(
   let js_runtime = JsRuntime::new(RuntimeOptions {
     will_snapshot: true,
+    modules: deno_modules,
     ..Default::default()
```
Very clean - I like it

ry commented on core/runtime_modules.rs:
```diff
@@ -0,0 +1,206 @@
+      $wrapper($opfn),
+    ),)+]
+  };
+}
```
I'm a bit wary of introducing macros. Although it reduces code, I worry that it obfuscates what's happening. That said, these seem relatively simple... Maybe consider doing these in a follow up PR.

bartlomieju replied:
I agree, I'm a bit wary of the `with()` middleware wrapper. This PR already introduces quite a few new concepts, I suggest to skip middleware implementation for a follow up PR.

AaronO replied:
@ry @bartlomieju What about something like:

```rust
    declare_ops!(json_op_async[
      op_webgpu_request_adapter,
      op_webgpu_request_device,
    ]),
```

That's a little easier to read, it conveys that `json_op_async` wraps `[...]` and hints that it returns a `Vec`

AaronO replied:
It could also be:

```rust
    declare_ops!(json_op_async, [
      op_webgpu_request_adapter,
      op_webgpu_request_device,
    ]),
```

If we want it to look more like a map expression (notice `, [`)

ry commented on core/runtime_modules.rs:
```diff
@@ -0,0 +1,206 @@
+    js_files: Vec<SourcePair>,
+    ops: Vec<OpPair>,
+    opstate_fn: Option<Box<OpStateFn>>,
+  ) -> Self {
```
Seems superfluous. Does this offer much more than `new()`? Using a couple fewer `Some` wrappers is offset by the mental overhead of having to understand all these different constructors.

AaronO replied:
@ry Not really, especially given that `vec![]` isn't terribly more verbose than `None` so `js_files` and `ops` could just be `Vec`s and not `Option`s in the struct.

I might introduce a builder pattern, something like:
```rust
BasicExtension::builder()
   .js(...)
   .ops(...)
   .state(move || { ... })
   .ops(...) // More ops to group by wrapper and "avoid" `declare_ops_group` ...
   .build()
```

ry commented on core/runtime_modules.rs:
```diff
@@ -0,0 +1,206 @@
+
+  pub fn pure_js(js_files: Vec<SourcePair>) -> Self {
+    Self::new(Some(js_files), None, None)
+  }
```
This one is nice

AaronO replied:
Thanks !

ry commented on core/runtime.rs:
```diff
@@ -81,6 +83,7 @@ pub struct JsRuntime {
   snapshot_creator: Option<v8::SnapshotCreator>,
   has_snapshotted: bool,
   allocations: IsolateAllocations,
+  modules: Vec<Box<dyn JsRuntimeModule>>,
```
I think we should use a different word than "module" because it might be confused with ES modules.

Currently we use the term "op_crates" but maybe we don't want to explicitly tie this to the rust crate concept. Maybe something like "op_package" ?

lucacasonato replied:
Maybe a RuntimeExtension?

AaronO replied:
Both are good candidates, I prefer `RuntimeExtension` but `OpBundle` isn't bad either. `OpBundle` is a little more specific and familiar given "op_crates", but `RuntimeExtension` is a better fit since runtime extensions can be pure JS too.

I could easily imagine the following in the docs:
> Runtime extensions implement the `deno_core::JsRuntimeExtension` interface(/trait), they extend the runtime's features by adding JS to the global scope and providing ops. Extensions are typically pure JS, a mix of JS & matching ops or stateful-ops & op-middleware in more advanced cases.

I'll continue with the current naming and reword on the "polish pass" once we've reached consensus.

ry replied:
RuntimeExtension sounds good to me

bartlomieju replied:
I'm strongly in favor of `op_crates`, this term is already prevalent

lucacasonato replied:
But not all op_crates contain ops 😉 (op_crates/web)

bartlomieju replied:
Sure... I just want to avoid situation where the same thing has two distinct names

AaronO replied:
`op_crates` isn't a bad name but it's a little too specific IMO:
- Extensions can be pure JS
- Extensions can be grouped (`runtime` will have many small op-extensions grouped with MultiExtensions)
- Plugins aren't op_crates but should return extensions (in the future)

There's no "perfect" candidate suggested so far, all have tradeoffs on familiarity/ambiguity/descriptiveness.

Obviously we could rename `op_crates` to `ext_crates` but that could be unnecessarily disruptive...

hayd replied:
Why not `Extension` (drop the `Runtime`)? and potentially op_crates->extensions later.

bartlomieju commented on core/runtime.rs:
```diff
@@ -349,6 +357,36 @@ impl JsRuntime {
+  // Inits JS of provided JsRuntimeModules
+  // NOTE: this will probably change when streamlining snapshot flow
+  pub fn init_mod_js(&mut self) -> Result<(), AnyError> {
+    let mut modules: Vec<Box<dyn JsRuntimeModule>> =
+      self.modules.drain(..).collect();
+    let selfish = Rc::new(RefCell::new(self));
+    for m in modules.iter_mut() {
+      let js_files = m.init_js()?;
+      for (filename, source) in js_files {
+        selfish.borrow_mut().execute_static(filename, source)?;
+      }
+    }
+    selfish.borrow_mut().modules = modules;
```
@AaronO could you explain the reason for "taking" modules, wrapping self in `Rc<RefCell<>>` and then putting modules back?

AaronO replied:
To avoid double-borrows, I've simplified it to remove the self wrap.

bartlomieju commented on core/runtime.rs:
```diff
@@ -180,6 +183,10 @@ pub struct RuntimeOptions {
 
+  /// JsRuntime modules, not to be confused with ES modules
+  /// these are deno modules containing JS + ops
+  pub modules: Vec<Box<dyn JsRuntimeModule>>,
```
These names should change as well - `JsRuntimeModule` cannot contain ES modules as source files, only plain old scripts, because currently snapshotting doesn't support ES modules.

AaronO replied:
Yes, that's pretty clear, it's linked to the above naming conversation, I'll change the names once there's some consensus.

rektide replied:
is there a v8 bug ticket for ES modules not being snapshottable? it would be nice if support was added. i searched v8 & chromium bug trackers but didn't turn up anything.

bartlomieju replied:
@rektide, not, that is not a V8 bug, it's lacking Deno implementation.

bartlomieju commented on core/runtime.rs:
```diff
@@ -411,6 +449,40 @@ impl JsRuntime {
   }
 
+  // same as execute() but for internal JS code
+  fn execute_static(
```
Seems superfluous, can't `execute` be used?

AaronO replied:
It can, but I wanted to experiment with https://github.com/denoland/deno/issues/9675, which introduces `execute_static` and passes those source strings as external v8 strings.

ry replied:
I want to experiment with that too, but I think first we need to support ExternalOneByteStringResource ExternalReferences. Let's remove this method for now?

AaronO replied:
Removed for now

bartlomieju commented on core/runtime_modules.rs:
```diff
@@ -0,0 +1,206 @@
+
+  fn init_ops(&mut self, registrar: RcOpRegistrar) -> Result<(), AnyError> {
+    // NOTE: not idempotent
+    // TODO: fail if called twice ?
```
Yes it should, I think `panic!` would be fine here

bartlomieju commented on core/runtime_modules.rs:
```diff
@@ -0,0 +1,206 @@
+////
+#[macro_export]
+macro_rules! include_js_files {
+  (root $root:literal, $($file:literal,)+) => {
```
How about renaming `root` to `prefix`? My first thought was that it tell a root path to resolve files, while actually it's the prefix that will be added to declared paths when executing files.

bartlomieju replied:
An example in the docstring would be helpful

AaronO replied:
@bartlomieju Are these good enough ? https://github.com/denoland/deno/blob/dcdd86c74eea328d6fdad78d5ffef7e65fb6d004/core/runtime_modules.rs#L242-L303

(also renamed `root` => `prefix`)

bartlomieju commented on op_crates/fetch/lib.rs:
```diff
@@ -409,30 +419,48 @@ where
+  } else {
+    None
+  };
+  create_http_client("".to_owned(), ca_data_vec)
```
This sets `User-Agent` to an empty string, which doesn't seem right. Why split into `create_http_client` two functions?

AaronO replied:
@bartlomieju There are already two duplicate`create_http_client` functions in `main`:
1. [runtime/http_util.rs](https://github.com/denoland/deno/blob/20627c91364d2a992fdfaaad7c8ae86454dbc2ed/runtime/http_util.rs#L13)
2. [op_crates/fetch/lib.rs](https://github.com/denoland/deno/blob/20627c91364d2a992fdfaaad7c8ae86454dbc2ed/op_crates/fetch/lib.rs#L412)

Both are broadly used for the fetch ops, the first is put into OpState the latter is used in `op_create_http_client`. So even in `main`, `op_create_http_client` doesn't allow setting the user-agent nor does it appear to use the "global" one from the OpState. `op_create_http_client` doesn't appear to be used anywhere so I was considering stripping it out or fixing it so it shares the user-agent.

lucacasonato replied:
@AaronO `op_create_http_client` is used for `Deno.createHttpClient`.

AaronO replied:
@lucacasonato Yep, sorry if I was unclear. I meant that I didn't see any calls to `Deno.createHttpClient` in our code or elsewhere (only [results on sourcegraph](https://sourcegraph.com/search?q=Deno.createHttpClient&patternType=literal) are deno's own tests).

But I would still argue that it's "incorrect" and should either accept a `userAgent` arg or default to the one provide to the default client.

lucacasonato replied:
It is indeed incorrect - the default user agent should be the same as the default http client. If it is not, that is a bug. Users can already specify a custom user agent by specifying the `User-Agent` header on requests.

AaronO replied:
I created an issue and provided a reproducible example here: https://github.com/denoland/deno/issues/9827, so it is confirmed, I'll fix it.

I do think being able to set `User-Agent` for a client would be convenient, though it can be worked around as you said, it's usually the kind of thing that is provided by the client/user-agent.

AaronO replied:
Fixed by https://github.com/denoland/deno/pull/9830

bartlomieju commented on runtime/build.rs:
```diff
@@ -44,8 +40,28 @@ fn create_snapshot(
 }
 
 fn create_runtime_snapshot(snapshot_path: &Path, files: Vec<PathBuf>) {
+  let deno_modules: Vec<Box<dyn JsRuntimeModule>> = vec![
+    Box::new(deno_webidl::init()),
+    Box::new(deno_console::init()),
+    Box::new(deno_url::init()),
+    Box::new(deno_web::init()),
+    Box::new(deno_fetch::init::<deno_fetch::NoFetchPermissions>(
+      "".to_owned(),
+      None,
+    )),
+    Box::new(
+      deno_websocket::init::<deno_websocket::NoWebSocketPermissions>(
+        "".to_owned(),
+        None,
+      ),
+    ),
+    Box::new(deno_crypto::init(None)),
+    Box::new(deno_webgpu::init(false)),
+  ];
```
I have some reservations about this - it requires to provide some state during snapshotting process which wasn't necessary before; it doesn't look so bad, but I can imagine situation with 20 crates and providing dummy values in such case might get complicated. Maybe there should be separate methods for obtaining JS sources and initializing the state?

AaronO replied:
This is indeed a small quirk and could be avoided, since JS provided by extensions isn't dependant on input args (currently and probably shouldn't). Both ops and opstate (and middleware once added) depend on args.

Splitting JS from ops would break the conceptual "bundle" that extensions provide.

Do you have a better idea on how to do this without breaking/leaking the extension abstraction ?

Currently I think dummy-args are a lesser evil, not ideal but better than a leaky abstraction.

Part of the goal of extensions, is to streamline snapshotting, so 3rd parties can easily create their own snapshots with relative ease.

AaronO replied:
@bartlomieju I think snapshotting JsRuntimes should be less of an edge case moving forward, in https://github.com/denoland/deno/pull/10308 I suggest allowing opcalls prior to snapshotting.

So rather than breaking the abstraction of runtime extensions for the sake of snapshots I think we should handle snapshots more gracefully and allow snapshotting fully initialized runtimes (with fully initialized extensions, js, ops, ...).

AaronO replied:
Snapshotting should no longer be an edge case since the op-layer rewrite and https://github.com/denoland/deno/pull/10308

AaronO commented:
@lucacasonato @bartlomieju I've implemented `OpMiddleware`:
- I added a [unit test](https://github.com/denoland/deno/blob/dcdd86c74eea328d6fdad78d5ffef7e65fb6d004/cli/tests/unit/metrics_test.ts#L70-L80) for it
- You can see the builder pattern in use in [metrics::init()](https://github.com/denoland/deno/blob/dcdd86c74eea328d6fdad78d5ffef7e65fb6d004/runtime/metrics.rs#L13-L24)
- Metrics is now passed as an extension in [worker.rs](https://github.com/denoland/deno/blob/dcdd86c74eea328d6fdad78d5ffef7e65fb6d004/runtime/worker.rs#L95-L96) and [web_worker.rs](https://github.com/denoland/deno/blob/dcdd86c74eea328d6fdad78d5ffef7e65fb6d004/runtime/web_worker.rs#L187-L188)
- So all the ops in the `op_crates` now go through this "middleware"

Middleware is important for this first iteration, since without it, op_crate ops' metrics would silently break (and it demonstrates that runtime extensions are general enough to cover our needs)

ry commented:
@AaronO I've done some renames and removed BasicModuleBuilder and MultiModule. We can add those in the future if necessary.

The Extension trait only has one implementor. Is it useful to even make it a trait? If we can express all op crates using the concrete BasicExtension struct, then maybe we should just rename that to Extension.

AaronO commented:
It appears we'll be landing a smaller scoped version of this (dropping the trait, `MultiModule` etc...)

For posteriority's sake here are the anticipated follow-ups:
1. Switch `deno_runtime` over to extensions
2. Maybe re-introduce `MultiModule` and the `Extension` trait (maybe re-introduce builder, etc...)
3. Switch plugins over to use this new declarative API
4. Use them to streamline snapshots and allow 3rd-parties to re-use `MainWorker` etc... with custom snapshots and custom runtime-extensions
5. An Unstable OpMiddleware to wrap ops that are fully unstable (NOT those that are conditionally/partially unstable)
6. A tracing OpMiddleware wrapping all ops to allow tracing similar to `strace` / `dtruss`

AaronO commented:
@ry https://github.com/denoland/deno/pull/9800/commits/ee50474397550566d40232048702a5e81af81bae downgrades the trait to the single concrete type.

Whilst it makes sense to drop the trait/interface in this limited scope, I'm fairly certain we'll have to restore it in the future

bartlomieju commented on core/extensions.rs:
```diff
@@ -0,0 +1,195 @@
+  /// Called at JsRuntime startup to initialize ops in the isolate.
+  pub fn init_ops(&mut self, registrar: RcOpRegistrar) -> Result<(), AnyError> {
+    // NOTE: not idempotent
+    // TODO: fail if called twice ?
+    if let Some(ops) = self.ops.take() {
+      for (name, opfn) in ops {
+        registrar.borrow_mut().register_op(name, opfn);
+      }
+    }
```
Again, I think it should panic if called second time

bartlomieju commented on core/extensions.rs:
````diff
@@ -0,0 +1,195 @@
+//   "01_hello.js",
+//   "02_goodbye.js",
+// )
+// ```
````
Should use triple slash to provide docstring

bartlomieju commented on core/extensions.rs:
````diff
@@ -0,0 +1,195 @@
+//    op_foo,
+//    op_bar,
+//  ]),
+// ```
````
Ditto

bartlomieju commented on core/extensions.rs:
````diff
@@ -0,0 +1,195 @@
+//    ...,
+//  ]),
+// ])
+// ```
````
Ditto

bartlomieju commented on core/lib.rs:
```diff
@@ -81,6 +83,11 @@ pub use crate::runtime::JsErrorCreateFn;
 pub use crate::runtime::RuntimeOptions;
 pub use crate::runtime::Snapshot;
+// pub use crate::runtime_modules::declare_ops!;
+// pub use crate::runtime_modules::include_js_files!;
```
Remove?

AaronO replied:
Yeah, I added those so when reviewing "exports" by checking `lib.rs` it's apparent that we export macros.

bartlomieju commented on op_crates/webgpu/lib.rs:
```diff
@@ -92,21 +99,23 @@ impl Resource for WebGPUQuerySet {
-  }
+    declare_webgpu_ops(),
+    Some(Box::new(move |state| {
+      // TODO: check & possibly streamline this
+      // Unstable might be able to be OpMiddleware
+      // let unstable_checker = state.borrow::<super::UnstableChecker>();
+      // let unstable = unstable_checker.unstable;
+      state.put(Unstable(unstable));
```
I don't like that this unstable checker is added in `webgpu` instead of in `runtime` crate

AaronO replied:
@bartlomieju `webgpu` (in `main`) already duplicates a good portion of the unstable code into it's own crate. We can remove this with an Unstable middleware in `runtime` (since I think none of `webgpu` ops are conditionally unstable)

bartlomieju commented on runtime/web_worker.rs:
```diff
@@ -165,11 +166,32 @@ impl WebWorker {
+      deno_crypto::init(options.seed),
+      deno_webgpu::init(options.unstable),
+      // Metrics
+      metrics::init(),
```
Not entirely sure about metrics being a middleware - I discussed it with @ry, concretely `metrics` do not properly track usage of zero copy buffers - eg. in case of "read" op metrics measure the size of buffer, but in reality it can only be filed with a few bytes (but the op will count the size of buffer instead). Maybe skip it for this PR?

AaronO replied:
@bartlomieju I agree that's an issue, but isn't that separate from it being middleware ?
(that issue exists on `main`)

Without that middleware, it will break metrics for op_crate ops (ops declared via this new extension system). I added a test that checks that metrics work for `op_url_parse` (op_crate metrics weren't tested at all, so it would have caused an unexpected fail/crash for users)

bartlomieju replied:
That's correct, but using metrics as a middleware I don't see a way how collection of metrics could be customized per op basis

lucacasonato replied:
@bartlomieju Metrics are exposed via `Deno.metrics` in stable, so we have to keep the current middleware collection until 2.0 at least.

AaronO replied:
@bartlomieju Was it ever customized before ? I might have missed that, I don't remember seeing anything to that effect.

It sounds like this is an orthogonal fix/change, that should happen in a follow-up PR and would have had to change with the new op-ABI anyway.

If you remove op-middleware and the metrics middleware right now, you'll cause breaking changes for users, unless you verbosely re-wrap all the ops but that kinda defeats the whole point of extension modularity.

ry commented on core/runtime.rs:
```diff
@@ -349,6 +358,41 @@ impl JsRuntime {
+    // Wrap registrar
+    for e in self.extensions.iter_mut() {
+      registrar = e.init_registrar(registrar.clone());
+    }
```
I'm confused about the flow of registrars and middleware.

Most of the Extensions do not register any middleware - EG op_crates/url - however somehow they all have Metrics applied to them? From what I can tell the init function in runtime/metrics.rs creates an Extension with middleware, and it is loaded as the last Extension for Worker/WebWorker. Does order matter in registering of extensions? It seems to be the case here.

AaronO replied:
@ry Order of extensions does NOT matter in regards to middleware, it basically inits all the middleware provided by all extensions before registering any of the ops.

bartlomieju replied:
I'm also confused what's going on here. Why does `init_registrar` returns another registrar? It seems that it happens for every registered extension. What's the use case?

AaronO replied:
It's essentially a reduce, transforming our initial OpRegistrar into a fully wrapped one.

We then register all ops using that fully wrapped OpRegistrar, so middleware provided by any of the extensions gets applied to all ops.

rtsmarty replied:
This seems like it could get a little hairy if you have more than a few middlewares. Each middleware is going to Box and RefCell its parent registrar (which is also Boxed and RefCelled). It's effectively a recursive queue of `Rc<RefCell<OpRegistrar>>`. Lots of needless pointer chasing and `borrow_mut`, as well as using a bunch of stack space for the recursive function calls.

Probably worth checking, but I'd guess that collecting the `Box<OpMiddlewareFn>` into a `Vec` and then making an `OpRegistrar` which iterates over and executes these `OpMiddlewareFn` before calling into the parent `OpRegistrar` will be faster and less allocation heavy. I also think it's easier to reason about but that's possibly just me :shrug: .

Something like this, which probably doesn't compile but hopefully you get the idea

```rust
impl Extension {
    // ...
    pub fn middleware_fn(&mut self) -> Option<Box<OpMiddlewareFn>> { /* ... */ }
    // ...
}

struct JsRuntimeOpRegistrar<'a> {
    middlewares: Vec<Box<OpMiddlewareFn>>,
    state: &'a mut OpState
}
impl<'a> OpRegistrar for JsRuntimeOpRegistrar<'a> {
    fn register_op(&mut self, name: &str, op_fn: Box<OpFn>) -> OpId {
        let mut new_op = op_fn;
        for middleware in &self.middlewares {
            new_op = middleware(name, new_op);
        }
        self.state.op_table.register_op(name, new_op);
    }
}

pub fn init_extension_ops(&mut self) -> Result<(), AnyError> {
    let op_state = self.op_state();
    let mut op_state = op_state.borrow_mut();
    let mut registrar = JsRuntimeOpRegistrar {
        middlewares: vec![],
        state: &mut *op_state,
    };
    for e in self.extensions.iter_mut() {
        if let Some(op) = e.middleware_fn() {
            registrar.middlewares.push(op);
        }
    }
    let registrar = Rc::new(RefCell::new(registrar));
    // ... continue as before
}
```

The same criticism can probably be applied to the `OpMiddlewareFn` effectively doing the same thing as well, I suppose, but that would probably require a much larger rewrite, from what I can see.

ry reviewed (commented):
@AaronO I've removed the `declare_ops` macro. There's nothing particularly wrong with something like that, but I think we can introduce it in a separate PR. I will leave the `include_js_files` because it's relatively straightforward and helpful.

AaronO commented:
@ry Are you in favour/ok with the `declare_ops!()` macro and simply want to "reduce scope" in this PR ? Or are you uncertain about it more broadly ?

I think it's genuinely useful and more-expressive, repeating the op_names as idents and strings has no upside and can get painfully verbose in `webgpu` to the point of hurting review-ability IMO. As always with these macros, they're designed to be opt-in, so users can just fallback to the "verbose version" if the macros don't fit their use-case (though it should work for most).

ry commented:
@AaronO Trying to reduce scope - but I'm a little concerned about the exact organization of them, particularly the nested `declare_ops_group`. I wonder if it wouldn't be better to just have a macro for json_op_sync and json_op_async to produce the OpPair. It's a discussion I'd rather punt on because it's rather superficial.

AaronO commented:
@ry Makes sense, happy to submit a follow-up PR and have a scoped conversation on that.

I had tested macros that just wrap an OpPair, but ops are often declared in "large" groups (definitely for `webgpu`) and so there's a good increase in expressiveness by wrapping multiple at once and also grouping them by type (sync vs async, etc...)

I had also tested non-parametric versions that take no wrapper, but those are less general as we'll preserve bin/buffer ops and start introducing min_ops (they could be aliased for the common cases, but that didn't seem like a substantial improvement to me)

bartlomieju commented on core/extensions.rs:
```diff
@@ -0,0 +1,136 @@
+  /// Called at JsRuntime startup to initialize ops in the isolate.
+  pub(crate) fn init_ops(&mut self, registrar: RcOpRegistrar) {
+    // NOTE: not idempotent
+    // TODO: fail if called twice ?
```
Yes, a panic would be fine, this really needs to be fixed before landing

AaronO replied:
Done

bartlomieju commented on core/extensions.rs:
```diff
@@ -0,0 +1,136 @@
+
+pub type SourcePair = (&'static str, &'static str);
+pub type OpPair = (&'static str, Box<OpFn>);
+pub type RcOpRegistrar = Rc<RefCell<dyn OpRegistrar>>;
```
I think I already mentioned it but IMO such alias adds even more indirection than using `Rc<RefCell<>>` directly

AaronO replied:
Dropped the alias

bartlomieju commented on core/runtime.rs:
```diff
@@ -351,6 +360,42 @@ impl JsRuntime {
   }
 
+  /// Initializes JS of provided Extensions
+  // NOTE: this will probably change when streamlining snapshot flow
```
Not sure what this comment means, remove before landing?

AaronO replied:
It's a note that `init_extension_js` and `init_extension_ops` should be fully merged into a `init_extensions()` method or even folded into `new` (with no `pub fn`s) as callers no longer need to handle them separately as snapshottable JsRuntimes call have full fledged ops.

bartlomieju commented on core/extensions.rs:
```diff
@@ -0,0 +1,136 @@
+impl Extension {
+  /// returns JS source code to be loaded into the isolate (either at snapshotting,
+  /// or at startup).  as a vector of a tuple of the file name, and the source code.
+  pub(crate) fn init_js(&self) -> Vec<SourcePair> {
```
This name is confusing, it doesn't init anything just returns the files right?

AaronO replied:
Extension was originally a trait and might become one again in the future, so in some sense how the JS files are obtained is opaque, also it makes sense to remain consistent with the other methods.

**Note:** these method names are used internally in core and aren't leaked to external users / op-crates

bartlomieju commented on core/extensions.rs:
```diff
@@ -0,0 +1,136 @@
+}
+
+impl Extension {
+  pub fn new(
+    js_files: Option<Vec<SourcePair>>,
+    ops: Option<Vec<OpPair>>,
+    opstate_fn: Option<Box<OpStateFn>>,
+    middleware_fn: Option<Box<OpMiddlewareFn>>,
+  ) -> Self {
```
Looking at the usages in various crates I'd suggest to use builder pattern instead. Once you got a few lines of js sources and a few elements in a vector of ops it becomes quite hard to read this call. WDYT?

AaronO replied:
I had implemented a builder pattern, Ryan stripped it out, I will restore it in a follow up PR

bartlomieju commented on core/runtime.rs:
```diff
@@ -351,6 +360,42 @@ impl JsRuntime {
+  }
+
+  /// Initializes ops of provided Extensions
+  // NOTE: this will probably change when streamlining snapshot flow
```
Ditto

AaronO replied:
Explained above (these methods shouldn't remain public mid term)

bartlomieju commented on runtime/web_worker.rs:
```diff
@@ -220,42 +247,24 @@ impl WebWorker {
         });
       }
 
+      js_runtime.init_extension_ops().unwrap();
```
Why unwrap?

AaronO replied:
Could change it to an `.expect()`, but ultimately we should panic here if it fails, the worker constructors don't return `Result`, though they obviously could. But I would rather do a 2nd pass and flesh out/polish error handling in Extensions.

bartlomieju reviewed (commented):
After more thought I really do think we shouldn't go with the name "extensions". A couple reasons:
- term "op crates" has been used for almost a year, it will be very confusing to transition to "extensions"
- "extension" is not very specific
- these are not "extensions" per se - they do not augment the capabilities of the runtime in any way, but rather allow to group ops inside a single crate

bartlomieju commented on core/runtime.rs:
```diff
@@ -351,6 +360,42 @@ impl JsRuntime {
+
+  /// Initializes ops of provided Extensions
+  // NOTE: this will probably change when streamlining snapshot flow
+  pub fn init_extension_ops(&mut self) -> Result<(), AnyError> {
```
I'm thinking that maybe this method should consume the `extensions` field? It shouldn't really be used after initializing the ops, right?

AaronO replied:
Plugins might be implemented as "dynamic extensions", in which case tracking them post init will streamline cleaning them up if we want to allow unloading plugins (which would make sense)

hayd commented on core/extensions.rs:
```diff
@@ -0,0 +1,136 @@
+impl Extension {
+  pub fn new(
+    js_files: Option<Vec<SourcePair>>,
+    ops: Option<Vec<OpPair>>,
```
What is the distinction between `None` and `Some(vec![])`  ?

AaronO replied:
Practically speaking there's no meaningful distinction, it's an implementation detail that will evolve and shouldn't be user facing but one upside of `Option` is that if the values need to be consumed there's a clean `.take()` operation

rtsmarty commented on core/extensions.rs:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  /// Called at JsRuntime startup to initialize ops in the isolate.
+  pub(crate) fn init_ops(&mut self, registrar: Rc<RefCell<dyn OpRegistrar>>) {
```
Is there a reason this is `Rc<RefCell<dyn OpRegistrar>>`? Looking at the function I feel it could probably just be a plain `&mut dyn OpRegistrar` and do exactly the same. If this eventually becomes a trait then it feels a little dodgy giving it an `Rc<RefCell<_>>`since it would allow extensions to stash the `OpRegistrar` away and use it to dynamically create ops during code execution (at least, in theory) which feels like an odd thing to support? I might be wrong, it could be completely sensible to want that.

AaronO replied:
I've killed the OpRegistrar trait and replaced it with first class op middleware.

rtsmarty commented on core/runtime.rs:
```diff
@@ -349,6 +357,45 @@ impl JsRuntime {
+  // NOTE: this will probably change when streamlining snapshot flow
+  pub fn init_extension_js(&mut self) -> Result<(), AnyError> {
+    // Take extensions to avoid double-borrow
+    let mut extensions: Vec<Extension> = self.extensions.drain(..).collect();
```
This will needlessly allocate & deallocate a second Vec for the extensions, you can use `std::mem::replace` to put an empty Vec (doesn't allocate) into its place instead and then replace it back once the call is done.

```rust
let mut extensions = std::mem::replace(&mut self.extensions, Vec::new()); // Allocation free
for m in extensions.iter_mut() {
    // ...
}
std::mem::replace(&mut self.extensions, extensions); // Put the extensions back.
```

AaronO replied:
Sure, that works too. Keep in mind that this code is very much not in a hot loop, so clarity matters more here than a small extra alloc (I'm fine with `std::mem::*` but it seems that it can mislead people to thinking that we're doing something low level and they need to pay careful attention)

bartlomieju reviewed (approved):
LGTM
